File: 1367.php
<?php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution {

    /**
     * @param ListNode $head
     * @param TreeNode $root
     * @return Boolean
     */
    function isSubPath($head, $root) {
        if (!$head) return true;
        if (!$root) return false;

        return $this->isSubPath($head, $root->left) || $this->isSubPath($head, $root->right) || $this->dfs($head, $root);
    }

    function dfs($head, $root) {
        if (!$head) return true;
        if (!$root) return false;

        if ($head->val !== $root->val) return false;

        return $this->dfs($head->next, $root->left) || $this->dfs($head->next, $root->right);
    }
}

File: go/linkedList/1367.go
package main

//type TreeNode struct {
//	Val   int
//	Left  *TreeNode
//	Right *TreeNode
//}
//
//type ListNode struct {
//	Val  int
//	Next *ListNode
//}

// 这种写法 当root.Val != head.Val时 head没有回到头节点
//func isSubPath(head *ListNode, root *TreeNode) bool {
//	if head == nil {
//		return true
//	}
//	if root == nil {
//		return false
//	}
//
//	if root.Val == head.Val {
//		return isSubPath(head.Next, root.Left) || isSubPath(head.Next, root.Right)
//	}
//	return isSubPath(head, root.Left) || isSubPath(head, root.Right)
//}

//func isSubPath(head *ListNode, root *TreeNode) bool {
//	if head == nil {
//		return true
//	}
//	if root == nil {
//		return false
//	}
//
//	return dfs(head, root) || isSubPath(head, root.Left) || isSubPath(head, root.Right)
//}
//
//func dfs(head *ListNode, root *TreeNode) bool {
//	if head == nil {
//		return true
//	}
//	if root == nil {
//		return false
//	}
//
//	if root.Val != head.Val {
//		return false
//	}
//	return dfs(head.Next, root.Left) || dfs(head.Next, root.Right)
//}